README.md
# bootComb

[![](https://cranlogs.r-pkg.org/badges/bootComb)](https://CRAN.R-project.org/package=bootComb)
[![](https://cranlogs.r-pkg.org/badges/grand-total/bootComb)](https://CRAN.R-project.org/package=bootComb)

Deriving confidence intervals for combinations of independent parameter estimates.

## Citation

Please use the following to cite the use of this R package in publications:

*  Henrion, M.Y.R. (2021). bootComb - An R Package to Derive Confidence Intervals for Combinations of Independent Parameter Estimates. International
  Journal of Epidemiology, dyab049. https://doi.org/10.1093/ije/dyab049

## Description

This is an R package to combine several estimates via a arbitrary function to compute a new parameter.

Uncertainty is propagated through by sampling from probability distributions for each estimate and then deriving an empirical interval containing 95% of the resulting distribution (either via highest density interval or quantiles).

Several specific applications, though the methodology is quite general:
1. Combining several conditional prevalences (see e.g. [Stockdale et al., J. Hepatol. (2020)](https://doi.org/10.1016/j.jhep.2020.04.008))
2. Adjusting a prevalence for assay sensitivity and specificity when the latter are not known exactly and only estimates are available.

By default, the method assumes that the combined estimates are independent. This is generally the case where parameters estimated on different, reasonably large datasets are combined, but in some situations this may not hold (e.g. when parameters themselves are not independent).
Where this assumption is likely not met, users can specify a correlation matrix for the input parameters. The package uses a Gaussian copula function to implement the dependence structure.

## Installation

### From CRAN (stable version)

``` r
install.packages('bootComb')
```

### From GitHub (development version)

``` r
# install.packages("devtools")
devtools::install_github("gitMarcH/bootComb")
```

## Example

95% confidence interval for the product of 2 prevalence parameters for which only the 95% confidence intervals are known. The `seed` argument is not needed - it just allows exact reproduction of the example.

``` r
library(bootComb)

dist1<-getBetaFromCI(qLow=0.2,qUpp=0.8,alpha=0.05)
dist2<-getBetaFromCI(qLow=0.4,qUpp=0.9,alpha=0.05)

distListEx<-list(dist1$r,dist2$r)
combFunEx<-function(pars){pars[[1]]*pars[[2]]}

bootComb(distList=distListEx,combFun=combFunEx,method="hdi",seed=123)
#> 
#> $conf.int
#>     lower     upper 
#> 0.1034161 0.5854474  
#> attr(,"credMass")
#> [1] 0.95
#> 
#> $bootstrapValues
#> NULL
```

Alternatively, the same result can also be arrived at in just 2 lines of code:

```r
combFunEx<-function(pars){pars[[1]]*pars[[2]]}
bootComb(distributions=c("beta","beta"),qLowVect=c(0.2,0.4),qUppVect=c(0.8,0.9),combFun=combFunEx,method="hdi",seed=123)
#> 
#> $conf.int
#>     lower     upper 
#> 0.1034161 0.5854474  
#> attr(,"credMass")
#> [1] 0.95
#> 
#> $bootstrapValues
#> NULL
```


NEWS.md
# bootComb 1.1.2

* Added the option to specify a dependence structure in `adjPrevSensSpecCI` (i.e. argument `Sigma` allowed and passed through to `bootComb`).
* Added the option to specify a seed value for `adjPrevSensSpecCI`.
* Added the option for bootstrapped values to be returned for `adjPrevSensSpecCI`.
* Added a note to the documentation files for getPoisFromCI and getNegBinFromCI - these two functions are unlikely to be of much use within bootComb (where parameters are necessarily assumed to be continuous).
* Minor fix to one example.

# bootComb 1.1.1

* Reduced the number of bootstrap samples for the example used for the `bootComb` example as CRAN gave an error that examples take over 10s to run.
* Added a seed value for the examples for the `bootComb` function.

# bootComb 1.1.0

* Support added for dependent / correlated parameters (using Gaussian copulas to specify the dependence). This is done by specifying a correlation matrix (new input argument `Sigma`).
* Added an alternative, and simpler, way of calling the main package function `bootComb` (new input arguments `distributions`, `qLowVect`, `qUppVect` and `alphaVect`).
* `returnBootVals` returns now also the values for the individual parameters that get combined. This change allows checking the dependence structure that was specified.
* A random seed can now be specified as an input argument to the `bootComb` function (argument `seed`).
* Update to README.md (example of alternative specification and citation info to avoid hard-coded reference to version number).


# bootComb 1.0.2

* Citation information updated now the IJE paper is out. Please cite https://doi.org/10.1093/ije/dyab049 when using the package.
* Updated the function `adjPrevSensSpecCI` for better behaviour in the case where the provided point estimates for the prevalence, sensitivity and specificity result in an impossible value (adjusted prevalence point estimate <0 or >1). The adjusted point estimates and boundaries of CI have now been updated to avoid impossible values and be consistent with each other (e.g. if the adjusted point estimate would be -0.000123, the lower boundary from the bootstrapping 0.000451 and the upper boundary 0.112345, then the reported point estimate and CI will be 0 [0,0.112345]).

# bootComb 1.0.1

* Fixed a bug with the internal functions `identifyBetaPars`, `identifyNormPars`, `identifyPoisPars`, `identifyNegBinPars`, `identifyGammaPars`, `identifyExpPars` whereby the `alpha` parameter did not get passed through.

# bootComb 1.0.0

With the inclusion of additional distributions functions that are supported, this is considered to be the first mature version of the package. Some changes for consistency of coding and better function naming are also minimally not backward compatible, hence the upgrade to version number 1.0.0.

* Added cranlogs badges to README.md
* Changed the way parameters are found for the Poisson distribution. Rather than quantile matching, the CDF is matched (as this will be a continuous function rather than a step function in the way the quantile function is).
* Added support for negative binomial, gamma and exonential distributions in addition to the beta, normal and Poisson distributions. Finding the best-fit negative binomial distribution proceeds similar to the procedure for the Poisson distribution as this is another discrete distribution and hence the quantile function is a step function (albeit a two-dimensional step function in this case).
* Renamed the (internal only) function 'ssPoisPar' to 'ssPoisPars' for consistency with other distributions.
* Renamed arguments 'pLow' and 'pUpp' from getBetaFromCI() to 'qLow' and 'qUpp' for consistency. Updated code from other functions where these parameters were specified.
* Default method for interval computation changed from `hdi` to `quantile` as the latter will always be correct and the former may be wrong if the bootstrap sample of combined parameter values is severely multimodal.
* Dependency on package `HDInterval` turned into a "Suggests" rather than an "Imports" dependency with `bootComb` falling back on `method='quantile'` if `HDInterval` is not available.
* Small update to the example from the README.md file.
* Clarified the names for the functions to run the 2 example simulations.

# bootComb 0.2.0

First version accepted & published on CRAN.

* Various bug fixes.
* Second simulation scenario added (for adjusted prevalence estimation situation).
* Ability to specify number of optimisation iterations for `identifyBetaPars`
* `identifyBetaPars` is no longer an exported function; strictly internal
* Added `getBetaFromCI`, a function that returns beta distribution functions from a given confidence interval (and in particular a random sampling function)
* Added support for normal and Poisson distributions in addition to the beta distribution
* Changed the organisation of code within the source R code files
* Implemented changes requested from the submission of v0.1.0 to CRAN (references in DESCRIPTION file, TRUE/FALSE instead of T/F, restoring user parameters after doing plots)

# bootComb 0.1.0

First version to be submitted to CRAN (but rejected).


R/simulations.R

#' @title Simulation scenario for the product of two prevlaence estimates.
#'
#' @description
#' This is a simulation to compute the coverage of the confidence interval returned by bootComb() in the case of the product of 2 probability parameter estimates.
#'
#' @param B The number of simulations to run. Defaults to 1e3.
#' @param p1 The true value of the first probability parameter.
#' @param p2 The true value of the second probability parameter.
#' @param nExp1 The size of each simulated experiment to estimate \code{p1}.
#' @param nExp2 TThe size of each simulated experiment to estimate \code{p2}.
#' @param alpha The confidence level; i.e. the desired coverage is 1-alpha. Defaults to 0.05.
#'
#' @return A list with 2 elements:
#'    \item{estimate}{A single number, the proportion of simulations for which the confidence interval contained the true parameter value.}
#'    \item{conf.int}{A 95\% confidence interval for the coverage estimate.}
#'
#' @examples
#' \donttest{
#' simScenProductTwoPrevs(p1=0.35,p2=0.2,nExp1=100,nExp2=1000,B=100)
#'   # B value only for convenience here
#'   # Increase B to 1e3 or 1e4 (be aware this may run for some time).
#'  }
#'
#' @export simScenProductTwoPrevs

simScenProductTwoPrevs<-function(B=1e3,p1,p2,nExp1,nExp2,alpha=0.05){
  trueP<-p1*p2
  res<-rep(0,B)

  for(j in 1:B){
    exp1<-rbinom(1,size=nExp1,prob=p1)
    exp2<-rbinom(1,size=nExp2,prob=p2)

    p1Est<-binom.test(x=exp1,n=nExp1)
    p2Est<-binom.test(x=exp2,n=nExp2)

    betaDist1<-getBetaFromCI(qLow=p1Est$conf.int[1],qUpp=p1Est$conf.int[2],initPars=c(round(nExp1/2),round(nExp1/2)))
    betaDist2<-getBetaFromCI(qLow=p2Est$conf.int[1],qUpp=p2Est$conf.int[2],initPars=c(round(nExp2/2),round(nExp2/2)))

    distListEx<-list(betaDist1$r,betaDist2$r)
    combFunEx<-function(pars){pars[[1]]*pars[[2]]}

    bsOut<-bootComb(distList=distListEx,combFun=combFunEx)

    if(trueP>=bsOut$conf.int[1] & trueP<=bsOut$conf.int[2]){res[j]<-1}
  }

  tmp<-binom.test(x=sum(res),n=length(res))
  coverage<-list(estimate=as.numeric(tmp$estimate),conf.int=tmp$conf.int) # ideally should be 95%

  return(coverage)
}

#' @title Simulation scenario for adjusting a prevalence for sensitivity and specificity.
#'
#' @description
#' This is a simulation to compute the coverage of the confidence interval returned by bootComb() in the case of adjusting a prevalence estimate for estimates of sensitivity and specificity.
#'
#' @param B The number of simulations to run. Defaults to 1e3.
#' @param p The true value of the prevalence parameter.
#' @param sens The true value of the assay sensitivity parameter.
#' @param spec The true value of the assay specificity parameter
#' @param nExp The size of each simulated experiment to estimate \code{p}.
#' @param nExpSens The size of each simulated experiment to estimate \code{sens}.
#' @param nExpSpec The size of each simulated experiment to estimate \code{spec}.
#' @param alpha The confidence level; i.e. the desired coverage is 1-alpha. Defaults to 0.05.
#' @param assumeSensSpecExact Logical; indicates whether coverage should also be computed for the situation where sensitivity and specificity are assumed to be known exactly. Defaults to FALSE.
#'
#' @return A list with 2 or 4 elements, depending whether \code{assumeSensSpecExact} is set to FALSE or TRUE:
#'    \item{estimate}{A single number, the proportion of simulations for which the confidence interval contained the true prevalence parameter value.}
#'    \item{conf.int}{A confidence interval of coverage 1-alpha for the coverage estimate.}
#'    \item{estimate.sensSpecExact}{Returned only if \code{assumeSensSpecExact} is set to TRUE. A single number, the proportion of simulations for which the confidence interval, derived assuming sensitivity and specificity are known exactly, contained the true prevalence parameter value.}
#'    \item{conf.int.sensSpecExact}{Returned only if \code{assumeSensSpecExact} is set to TRUE. A confidence interval of coverage 1-alpha for the coverage estimate in the scenario where sensitivity and specificity are assumed to be known exactly.}
#'
#' @examples
#' \donttest{
#' simScenPrevSensSpec(p=0.15,sens=0.85,spec=0.90,nExp=300,nExpSens=600,nExpSpec=400,B=100)
#'   # B value only for convenience here
#'   # Increase B to 1e3 or 1e4 (be aware this may run for some time).
#'  }
#'
#' @export simScenPrevSensSpec

simScenPrevSensSpec<-function(B=1e3,p,sens,spec,nExp,nExpSens,nExpSpec,alpha=0.05,assumeSensSpecExact=FALSE){
  trueObsPrev<-p*sens+(1-p)*(1-spec)
  res<-rep(0,B)
  if(assumeSensSpecExact){res2<-rep(0,B)}

  for(j in 1:B){
    expP<-rbinom(1,size=nExp,prob=trueObsPrev)
    expSens<-rbinom(1,size=nExpSens,prob=sens)
    expSpec<-rbinom(1,size=nExpSpec,prob=spec)

    pEst<-binom.test(x=expP,n=nExp)
    sensEst<-binom.test(x=expSens,n=nExpSens)
    specEst<-binom.test(x=expSpec,n=nExpSpec)

    bsOut<-adjPrevSensSpecCI(prevCI=pEst$conf.int,sensCI=sensEst$conf.int,specCI=specEst$conf.int,alpha=alpha,method="hdi")

    if(p>=bsOut$conf.int[1] & p<=bsOut$conf.int[2]){res[j]<-1}

    if(assumeSensSpecExact){
      ciTmp<-c((pEst$conf.int[1]-(1-specEst$estimate))/(sensEst$estimate-(1-specEst$estimate)),(pEst$conf.int[2]-(1-specEst$estimate))/(sensEst$estimate-(1-specEst$estimate)))
      if(p>=ciTmp[1] & p<=ciTmp[2]){res2[j]<-1}
    }
  }

  tmp<-binom.test(x=sum(res),n=length(res))
  coverage<-list(estimate=as.numeric(tmp$estimate),conf.int=tmp$conf.int) # ideally should be 95%

  if(assumeSensSpecExact){
    tmp<-binom.test(x=sum(res2),n=length(res2))
    coverage$estimate.sensSpecExact<-as.numeric(tmp$estimate)
    coverage$conf.int.sensSpecExact<-tmp$conf.int
  }

  return(coverage)
}


R/bootComb.R
#' @title Combine parameter estimates via bootstrap
#'
#' @description
#' This package propagates uncertainty from several estimates when combining these estimates via a function.
#' It does this by using the parametric bootstrap to simulate values from the distribution of each estimate to build up an empirical distribution of the combined parameter.
#' Finally either the percentile method is used or the highest density interval is chosen to derive a confidence interval for the combined parameter with the desired coverage.
#'
#' @param distList If \code{Sigma} is set to NULL, this is a list object where each element of the list is a sampling function for a probability distribution function (i.e. like rnorm, rbeta, ...). If \code{Sigma} is specified, then this needs to be a list of quantile functions for the distributions for each parameter.
#' @param combFun The function to combine the different estimates to a new parameter. Needs to take a single list as input argument, one element of the list for each estimate. This list input argument needs to be a list of same length as distList.
#' @param N The number of bootstrap samples to take. Defaults to 1e6.
#' @param distributions Alternatively to specifying \code{distlist}, the parameters \code{distributions}, \code{qLowVect}, \code{qUppVect} and (optionally) \code{alphaVect} can be specified. The first 3 of these need t be either all specified and be vectors of the same length or all set to NULL. The \code{distributions} parameter needs to be a vector specifying the names of the distributions for each parameter (one of "beta", "exponential", "gamma", "normal", "Poisson" or "NegativeBinomial").
#' @param qLowVect Alternatively to specifying \code{distlist}, the parameters \code{distributions}, \code{qLowVect}, \code{qUppVect} and (optionally) \code{alphaVect} can be specified. The first 3 of these need t be either all specified and be vectors of the same length or all set to NULL. The \code{qLowVect} parameter needs to be a vector specifying the lower confidence interval limits for each parameter.
#' @param qUppVect Alternatively to specifying \code{distlist}, the parameters \code{distributions}, \code{qLowVect}, \code{qUppVect} and (optionally) \code{alphaVect} can be specified. The first 3 of these need t be either all specified and be vectors of the same length or all set to NULL. The \code{qUppVect} parameter needs to be a vector specifying the upper confidence interval limits for each parameter.
#' @param alphaVect Alternatively to specifying \code{distlist}, the parameters \code{distributions}, \code{qLowVect}, \code{qUppVect} and (optionally) \code{alphaVect} can be specified. The first 3 of these need t be either all specified and be vectors of the same length or all set to NULL. The \code{alphaVect} parameter needs to be a vector specifying the alpha level (i.e. 1 minus the coverage) of each confidence interval. Can be specified as a single number if the same for all parameters. Defaults to 0.05.
#' @param Sigma Set to NULL if parameters are assumed to be independent (the default). If specified, this needs to be a valid covariance matrix for a multivariate normal distribution with variances equal to 1 for all variables (in other words, this really is a correlation matrix).
#' @param method The method uses to derive a confidence interval from the empirical distribution of the combined parameter.Needs to be one of 'quantile' (default; uses the percentile method to derive the confidence interval) or hdi' (computes the highest density interval).
#' @param coverage The desired coverage of the resulting confidence interval.Defaults to 0.95.
#' @param doPlot Logical; indicates whether a graph should be produced showing the input distributions and the resulting empirical distribution of the combined estimate together with the reported confidence interval. Defaults to FALSE.
#' @param legPos Legend position (only used if doPlot==TRUE); either NULL (no legend) or one of "top", "topleft", "topright", "bottom", "bottomleft", "bottomright" "left", "right", "center".
#' @param returnBootVals Logical; if TRUE then the parameter values computed from the bootstrapped input parameter values will be returned; values for the individual parameters will be reported as a second list element; defaults to FALSE.
#' @param validRange Optional; if not NULL, a vector of length 2 giving the range within which the values obtained from the bootstrapped input parameters must lie; values outside this range will be discarded. Behaviour that results in the need for this option arises when parameters are not independent. Use with caution.
#' @param seed If desired a random seed can be specified so that the same results can be reproduced.
#'
#' @return A list with 3 elements:
#' \item{conf.int}{A vector of length 2 giving the lower and upper limits of the computed confidence interval.}
#' \item{bootstrapValues}{A vector containing the computed / combined parameter values from the bootstrap samples of the input parameters. (Only non-NULL if \code{returnBootVals} is set to TRUE.)}
#' \item{bootstrapValuesInput}{A list where each element is the vector of the bootstrapped values for the corresponding input parameter. This can be useful to check the dependence structure that was specified. (Only non-NULL if \code{returnBootVals} is set to TRUE.)}
#'
#' @seealso
#' \code{\link[HDInterval]{hdi}}
#'
#' @examples
#' ## Example 1 - product of 2 probability parameters for which only the 95% CIs are reported
#' dist1<-getBetaFromCI(qLow=0.4,qUpp=0.6,alpha=0.05)
#' dist2<-getBetaFromCI(qLow=0.7,qUpp=0.9,alpha=0.05)
#' distListEx<-list(dist1$r,dist2$r)
#' combFunEx<-function(pars){pars[[1]]*pars[[2]]}
#' bootComb(distList=distListEx,
#'          combFun=combFunEx,
#'          doPlot=TRUE,
#'          method="hdi",
#'          N=1e5, # reduced from N=1e6 so that it runs quicker; larger values => more accurate
#'          seed=352)
#'
#' # Alternatively, the same example can be run in just 2 lines of code:
#' combFunEx<-function(pars){pars[[1]]*pars[[2]]}
#' bootComb(distributions=c("beta","beta"),
#'          qLowVect=c(0.4,0.7),
#'          qUppVect=c(0.6,0.9),
#'          combFun=combFunEx,
#'          doPlot=TRUE,
#'          method="hdi",
#'          N=1e5, # reduced from N=1e6 so that it runs quicker; larger values => more accurate
#'          seed=352)
#'
#' ## Example 2 - sum of 3 Gaussian distributions
#' dist1<-function(n){rnorm(n,mean=5,sd=3)}
#' dist2<-function(n){rnorm(n,mean=2,sd=2)}
#' dist3<-function(n){rnorm(n,mean=1,sd=0.5)}
#' distListEx<-list(dist1,dist2,dist3)
#' combFunEx<-function(pars){pars[[1]]+pars[[2]]+pars[[3]]}
#' bootComb(distList=distListEx,combFun=combFunEx,doPlot=TRUE,method="quantile")
#'
#' # Compare with theoretical result:
#' exactCI<-qnorm(c(0.025,0.975),mean=5+2+1,sd=sqrt(3^2+2^2+0.5^2))
#' print(exactCI)
#' x<-seq(-10,30,length=1e3)
#' y<-dnorm(x,mean=5+2+1,sd=sqrt(3^2+2^2+0.5^2))
#' lines(x,y,col="red")
#' abline(v=exactCI[1],col="red",lty=3)
#' abline(v=exactCI[2],col="red",lty=3)
#'
#' ## Example 3 - same as Example 1 but assuming the 2 parameters to be dependent / correlated
#' combFunEx<-function(pars){pars[[1]]*pars[[2]]}
#' bootComb(distributions=c("beta","beta"),
#'          qLowVect=c(0.4,0.7),
#'          qUppVect=c(0.6,0.9),
#'          Sigma=matrix(byrow=TRUE,ncol=2,c(1,0.5,0.5,1)),
#'          combFun=combFunEx,
#'          doPlot=TRUE,
#'          method="hdi",
#'          N=1e5, # reduced from N=1e6 so that it runs quicker; larger values => more accurate
#'          seed=352)
#'
#' @export bootComb

bootComb<-function(distList,combFun,N=1e6,distributions=NULL,qLowVect=NULL,qUppVect=NULL,alphaVect=0.05,Sigma=NULL,method="quantile",coverage=0.95,doPlot=FALSE,legPos="topright",returnBootVals=FALSE,validRange=NULL,seed=NULL){
  if(!is.null(distributions) & !is.null(qLowVect) & !is.null(qUppVect)){
    K<-length(distributions)
    if(K!=length(qLowVect) | K!=length(qUppVect)){
      stop("Arguments 'distributions', 'qLowVect' and 'qUppVect' need to be vectors of the same length.")
    }
    if(length(alphaVect)!=1 & length(alphaVect)!=K){
      stop("Argument 'alphaVect' needs to be either a vector of length 1 (a single number) or a vector of same length as arguments 'distributions', 'qLowVect' and 'qUppVect'.")
    }
    if(sum(alphaVect<0 | alphaVect>1)>0){
      stop("All elements of argument 'alphaVect' need to be values between 0 and 1.")
    }
    if(length(alphaVect)==1){alphaVect<-rep(alphaVect,K)}

    distList<-vector("list",K)

    for(k in 1:K){
      if(tolower(distributions[k])=="beta"){distTmp<-getBetaFromCI(qLow=qLowVect[k],qUpp=qUppVect[k],alpha=alphaVect[k])}
      if(tolower(distributions[k])=="exponential"){distTmp<-getExpFromCI(qLow=qLowVect[k],qUpp=qUppVect[k],alpha=alphaVect[k])}
      if(tolower(distributions[k])=="gamma"){distTmp<-getGammaFromCI(qLow=qLowVect[k],qUpp=qUppVect[k],alpha=alphaVect[k])}
      if(tolower(distributions[k])=="normal"){distTmp<-getNormFromCI(qLow=qLowVect[k],qUpp=qUppVect[k],alpha=alphaVect[k])}
      if(tolower(distributions[k])=="poisson"){distTmp<-getPoisFromCI(qLow=qLowVect[k],qUpp=qUppVect[k],alpha=alphaVect[k])}
      if(tolower(distributions[k])=="negativebinomial"){distTmp<-getNegBinFromCI(qLow=qLowVect[k],qUpp=qUppVect[k],alpha=alphaVect[k])}

      if(is.null(Sigma)){
        distList[[k]]<-distTmp$r
      }else{
        distList[[k]]<-distTmp$q
      }
    }
  }else if(!is.null(distributions) | !is.null(qLowVect) | !is.null(qUppVect)){
    stop("Either all of the arguments 'distributions', 'qLowVect' and 'qUppVect' have to be specified or all have to be set to NULL.")
  }

  pars<-vector("list",length=length(distList))

  if(!is.null(seed)){set.seed(seed)}

  if(is.null(Sigma)){
    for(k in 1:length(distList)){
      pars[[k]]<-distList[[k]](N)
    }
  }else{
    if(sum(diag(Sigma)!=rep(1,nrow(Sigma)))>0){stop("All diagonal elements of the Sigma matrix need to be equal to 1.")}
    parsNorm<-MASS::mvrnorm(n=N,mu=rep(0,length(distList)),Sigma=Sigma)
    parsNorm<-pnorm(parsNorm)
    for(k in 1:length(distList)){
      pars[[k]]<-distList[[k]](parsNorm[,k])
    }
    rm(parsNorm)
  }

  combParVals<-combFun(pars)
  if(!is.null(validRange)){
    combParVals<-combParVals[combParVals>=validRange[1] & combParVals<=validRange[2]]
    if(length(combParVals)==0){stop("All sample values are outside the valid range.")}
  }

  if(method=="hdi"){
    if(requireNamespace("HDInterval",quietly=TRUE)){
      ci<-HDInterval::hdi(combParVals,credMass=coverage)
    }else{
      warning("The bootComb argument 'method' was set to 'hdi', but the required R package 'HDInterval is not installed. Falling back on method='quantile'.")
      ci<-quantile(combParVals,probs=c((1-coverage)/2,(1+coverage)/2))
    }
  }else if(method=="quantile"){
    ci<-quantile(combParVals,probs=c((1-coverage)/2,(1+coverage)/2))
  }else{
    stop(paste(sep=" ","The argument 'method' to function 'bootComb' needs to be one of 'hdi', 'quantile' (received <",method,">)."))
  }

  if(doPlot){
    oldpar <- par(no.readonly = TRUE)
    on.exit(par(oldpar))

    layout(matrix(c(1:length(distList),rep(length(distList)+1,length(distList))),nrow=2,byrow=TRUE))

    for(k in 1:length(distList)){
      hist(pars[[k]],breaks=100,xlab=names(distList)[k],ylab="density",freq=FALSE,main=paste(sep="","Histogram - sampled values for parameter ",k,"."))
    }

    hist(combParVals,breaks=100,freq=FALSE,xlab="combined parameter",ylab="density",main="Empirical distribution of the combined parameter.")
    abline(v=ci[1],lty=2,lwd=2,col="steelblue")
    abline(v=ci[2],lty=2,lwd=2,col="steelblue")
    if(!is.null(legPos)){
      legend(x=legPos,bty="n",horiz=TRUE,lwd=2,lty=2,col=c("steelblue"),legend=c("95% confidence limits"))
    }
  }

  if(!returnBootVals){
    combParVals<-NULL
    pars<-NULL
  }

  return(list(conf.int=ci,bootstrapValues=combParVals,bootstrapValuesInput=pars))
}
